ast/desugar/Desugar.cc

```diff
@@ -404,6 +404,41 @@ TreePtr desugarMlhs(DesugarContext dctx, core::LocOffsets loc, parser::Mlhs *lhs
     return MK::InsSeq(loc, std::move(stats), MK::Local(loc, tempRhs));
 }
 
+// Map all MatchVars used in `pattern` to local variables initialized from magic calls
+void desugarPatternMatchingVars(InsSeq::STATS_store &vars, DesugarContext dctx, unique_ptr<parser::Node> &node) {
+    if (auto var = parser::cast_node<parser::MatchVar>(node.get())) {
+        auto loc = var->loc;
+        auto recv = MK::Constant(loc, core::Symbols::Magic());
+        auto val = MK::Send0(loc, std::move(recv), core::Names::patternMatchVar());
+        vars.emplace_back(MK::Assign(loc, var->name, std::move(val)));
+    } else if (auto rest = parser::cast_node<parser::MatchRest>(node.get())) {
+        desugarPatternMatchingVars(vars, dctx, rest->var);
+    } else if (auto pair = parser::cast_node<parser::Pair>(node.get())) {
+        desugarPatternMatchingVars(vars, dctx, pair->value);
+    } else if (auto as_pattern = parser::cast_node<parser::MatchAs>(node.get())) {
+        auto name = parser::cast_node<parser::MatchVar>(as_pattern->as.get())->name;
+        auto recv = MK::Constant(as_pattern->as->loc, core::Symbols::Magic());
+        auto val = MK::Send0(as_pattern->as->loc, std::move(recv), core::Names::patternMatchAs());
```

Comment:
What is the difference between `<Magic>.<pattern-match-var>` and `<Magic>.<pattern-match-as>`?

Could we make them both use `ast::MK::Unimplemented`?

Reply:
`match-var` was meant to access a matched variable like `a`, while `match-as` was to access a bound variable like `a => Integer`.

Replaced both by `RaiseUnimplemented`.